nomad/csi_endpoint.go

```diff
@@ -295,6 +295,28 @@ func (v *CSIVolume) Claim(args *structs.CSIVolumeClaimRequest, reply *structs.CS
 		return structs.ErrPermissionDenied
 	}
 
+	// Here we validate the existence of the volume and the allocation early to
+	// provide better errors to clients and to ensure that all the requisite state
+	// is available before starting to perform external actions.
+	state := v.srv.fsm.State()
+	ws := memdb.NewWatchSet()
+	vol, err := state.CSIVolumeByID(ws, args.VolumeID)
```

Comment:
We already do `CSIVolumeByID` as a validation step in `controllerPublishVolume` (because we have to have the volume to determine whether or not we do the controller publish/unpublish step). But looking at this now, that seems kind of by-accident rather an explicit validation step.

But rather than changing this, let's leave it in place and then I've opened https://github.com/hashicorp/nomad/issues/7175 to fix the logic in `controllerPublishVolume`